### PR TITLE
feat: add present proof v2

### PR DESCRIPTION
**Pull request by NB-MikeRichardson**
This is essentially a three-way merge of

feat/ppv2
0.3.0-pre
TimoGlastra/refactor/generalize-proofs

BREAKING CHANGE: the present proof module has been extended to support both v1 and v2 of protocol. See [Migrating from AFJ 0.2.x to 0.3.0](https://github.com/hyperledger/aries-framework-javascript/blob/main/docs/migration/0.2-to-0.3.md) for a detailed migration instructions.

### CODE REVIEW

**Comment by TimoGlastra** [comment]
@NB-MikeRichardson it seems one test is failing, could you take a look at it?

**Review by genaris — COMMENTED** [review]
I've just added some minor comments regarding RevocationNotificationService that I think belongs to Credential Module. 

It would be nice if we can integrate here the changes introduced in #972 but if it's too much trouble we can merge it afterwards.

File: packages/core/src/modules/proofs/formats/ProofFormat.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * }
+ * ```
+ */
+export type ProofFormatPayload<PFs extends ProofFormat[], M extends keyof ProofFormat['proofFormats']> = {
+  [ProofFormat in PFs[number] as ProofFormat['formatKey']]?: ProofFormat['proofFormats'][M]
+}
+
+export interface ProofFormat {
+  formatKey: string // e.g. 'ProofManifest', cannot be shared between different formats
+  proofRecordType: string // e.g. 'w3c', can be shared between multiple formats
```
[the remark this answers is not in the document]
Comment (NB-MikeRichardson) [reply packages/core/src/modules/proofs/formats/ProofFormat.ts:26]: I've taken this out. Will we need it later to distinguish indy vs presentation-exchange formats?

File: packages/core/src/modules/proofs/models/ProofServiceOptions.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import type { ConnectionRecord } from '../../connections'
+import type { ProofFormat, ProofFormatPayload } from '../formats/ProofFormat'
+import type { IndyProofFormat } from '../formats/indy/IndyProofFormat'
```
[the remark this answers is not in the document]
Comment (NB-MikeRichardson) [reply packages/core/src/modules/proofs/models/ProofServiceOptions.ts:3]: Ran yarn lint and removed all 'unused' warnings

**Comment by NB-MikeRichardson** [comment]
I haven't merged in https://github.com/hyperledger/aries-framework-javascript/pull/972 because that PR is not yet in 0.3.0-pre.

I am happy to merge that it in in due course: I would say let's get this stuff in first as people are waiting on it.

File: packages/core/src/modules/proofs/ProofsModule.ts [hunk 3]
```diff
@@ -1,32 +1,30 @@
...
+import { IndyProofFormatService } from './formats/indy/IndyProofFormatService'
+import { V1ProofService } from './protocol/v1'
+import { V2ProofService } from './protocol/v2'
 import { ProofRepository } from './repository'
-import { ProofService } from './services'
 
 export class ProofsModule implements Module {
-  public readonly config: ProofsModuleConfig
-
-  public constructor(config?: ProofsModuleConfigOptions) {
```
Comment (TimoGlastra) [inline packages/core/src/modules/proofs/ProofsModule.ts:12]: We should keep the config

Comment (NB-MikeRichardson) [reply packages/core/src/modules/proofs/ProofsModule.ts]: I assume it needs the registerInstance call

**Comment by TimoGlastra** [comment]
> It would be nice if we can integrate here the changes introduced in https://github.com/hyperledger/aries-framework-javascript/pull/972 but if it's too much trouble we can merge it afterwards.

Yes I think it would be easier to merge afterwards. I'll merge this PR and we can then update with main (hasn't been done in a while).

**Comment by TimoGlastra** [comment]
Did someone add sonarcloud?

**Review by TimoGlastra — APPROVED** [review]

**Comment by TimoGlastra** [comment]
Merging as CI for node 14 passes (still the BBS issue)

**Pull request merged**